lib/evaluate/characteristics.h

```diff
@@ -95,8 +104,9 @@ class TypeAndShape {
   const Attrs &attrs() const { return attrs_; }
 
   int Rank() const { return GetRank(shape_); }
-  bool IsCompatibleWith(
-      parser::ContextualMessages &, const TypeAndShape &) const;
+  bool IsCompatibleWith(parser::ContextualMessages &, const TypeAndShape &,
+      const char * = "POINTER", const char * = "TARGET",
+      bool isElemental = false) const;
```

Comment:
`isElemental` might deserve a comment. Looking at the signature, I thought it was a way to ask the function to accept different shapes as long as one of the shape is a scalar, but looking at the way it is implemented and called, I now understand that it actually means that shape conformance test should be skipped, which means the caller is responsible to ensure that one of the shape is scalar when setting this flag.

Reply:
But if the reader is not familiar with `ELEMENTAL` procedures in Fortran, this may not be the best place to describe them.  What it is supposed to do here is skip the shape conformance test between an actual argument and a dummy, since an `ELEMENTAL` procedure's dummy arguments are required to be scalar, but the actual arguments are allowed to be arrays (and must be conformant with each other).